juju/juju
# Ensure k8s resources are removed when an application is removed

## Description of change

Main fix here is to ensure k8s resources are cleaned up when an application is removed from Juju. Such resources include units, service, operator. As a driveby, toggle the default deployment strategy to use a deployment controller. A lot of this is api facade boilerplate - the caas operator facade gets a Life() method added.

A few things were preventing cleanup. The firewaller was exiting with an error on app not found. The responsibility for deleting units has been shifted to the main application worker. We now remember to delete the ingress controller if an application is exposed. And code is added to remove the operator.

## QA steps

Deploy a caas charm, add units, expose application.
Remove application and ensure everything is deleted from the k8s cluster.
